File: packages/core-go/cid.go
package arky

import (
	"crypto/sha256"
	"errors"
	"strings"
)

// base58btc (Bitcoin) alphabet.
const b58Alphabet = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"

// Base58BtcEncode encodes bytes as base58btc. Hand-rolled: the standard library
// has no base58, and the algorithm is short enough that a dependency would cost
// more trust than it saves.
func Base58BtcEncode(input []byte) string {
	if len(input) == 0 {
		return ""
	}
	// Leading zero bytes each map to a literal '1'.
	zeros := 0
	for zeros < len(input) && input[zeros] == 0 {
		zeros++
	}
	// Repeated division of the big-endian number by 58.
	buf := make([]byte, len(input))
	copy(buf, input)
	var out []byte
	for start := zeros; start < len(buf); {
		carry := 0
		for i := start; i < len(buf); i++ {
			cur := carry*256 + int(buf[i])
			buf[i] = byte(cur / 58)
			carry = cur % 58
		}
		out = append(out, b58Alphabet[carry])
		for start < len(buf) && buf[start] == 0 {
			start++
		}
	}
	var sb strings.Builder
	sb.Grow(zeros + len(out))
	for range zeros {
		sb.WriteByte('1')
	}
	for i := len(out) - 1; i >= 0; i-- {
		sb.WriteByte(out[i])
	}
	return sb.String()
}

// Base58BtcDecode decodes a base58btc string. It returns an error for any
// character outside the alphabet, so malformed input can never be mistaken for
// a valid key or digest.
func Base58BtcDecode(s string) ([]byte, error) {
	if s == "" {
		return []byte{}, nil
	}
	zeros := 0
	for zeros < len(s) && s[zeros] == '1' {
		zeros++
	}
	buf := make([]byte, 0, len(s))
	for i := zeros; i < len(s); i++ {
		idx := strings.IndexByte(b58Alphabet, s[i])
		if idx < 0 {
			return nil, errors.New("base58: invalid character")
		}
		carry := idx
		for j := len(buf) - 1; j >= 0; j-- {
			cur := int(buf[j])*58 + carry
			buf[j] = byte(cur % 256)
			carry = cur / 256
		}
		for carry > 0 {
			buf = append([]byte{byte(carry % 256)}, buf...)
			carry /= 256
		}
	}
	out := make([]byte, zeros+len(buf))
	copy(out[zeros:], buf)
	return out, nil
}

// MultihashSha256 returns multihash(sha2-256, data) = 0x12 0x20 || digest.
func MultihashSha256(data []byte) []byte {
	sum := sha256.Sum256(data)
	mh := make([]byte, 0, 34)
	mh = append(mh, 0x12, 0x20)
	return append(mh, sum[:]...)
}

// ToMultibase encodes bytes as multibase base58btc (the 'z' prefix).
func ToMultibase(data []byte) string { return "z" + Base58BtcEncode(data) }

// FromMultibase decodes a multibase 'z...' (base58btc) string.
func FromMultibase(s string) ([]byte, error) {
	rest, ok := strings.CutPrefix(s, "z")
	if !ok {
		return nil, errors.New("multibase: expected 'z' prefix")
	}
	return Base58BtcDecode(rest)
}

// MultihashMb returns multibase(multihash(sha2-256, data)).
func MultihashMb(data []byte) string { return ToMultibase(MultihashSha256(data)) }

// CidFromCanonical computes the cid over canonical bytes, per
// ARKY-TIM-Canonicalization-v1 section 4:
//
//	cid = multibase('z', base58btc(multihash(sha2-256, canonical_bytes)))
//
// This is NOT an IPFS CID.
func CidFromCanonical(canonical string) string {
	return MultihashMb([]byte(canonical))
}

// ComputeCid canonicalizes a value and returns its cid.
func ComputeCid(v Value) (string, error) {
	canonical, err := Canonicalize(v)
	if err != nil {
		return "", err
	}
	return CidFromCanonical(canonical), nil
}

File: packages/core-py/src/arky_core/cid.py
"""Content addressing per ARKY-TIM-Canonicalization-v1 section 4.

``cid = multibase('z', base58btc(multihash(sha2-256, canonical_bytes)))``,
where ``multihash = 0x12 0x20 || sha256(bytes)``. This is NOT an IPFS CID.
"""

from __future__ import annotations

import hashlib

from .canonicalize import canonicalize

__all__ = [
    "Base58Error",
    "base58btc_decode",
    "base58btc_encode",
    "cid_from_canonical",
    "compute_cid",
    "from_multibase",
    "multihash_mb",
    "multihash_sha256",
    "to_multibase",
]

_ALPHABET = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"
_INDEX = {ch: i for i, ch in enumerate(_ALPHABET)}


class Base58Error(ValueError):
    """Raised for input outside the base58btc alphabet or multibase prefix."""


def base58btc_encode(data: bytes) -> str:
    """Encode bytes as base58btc (hand-rolled; there is no stdlib base58)."""
    if not data:
        return ""
    zeros = 0
    while zeros < len(data) and data[zeros] == 0:
        zeros += 1
    num = int.from_bytes(data, "big")
    out: list[str] = []
    while num > 0:
        num, rem = divmod(num, 58)
        out.append(_ALPHABET[rem])
    # Each leading zero byte encodes as a literal '1'.
    return "1" * zeros + "".join(reversed(out))


def base58btc_decode(text: str) -> bytes:
    """Decode a base58btc string, raising :class:`Base58Error` on bad input."""
    if text == "":
        return b""
    zeros = 0
    while zeros < len(text) and text[zeros] == "1":
        zeros += 1
    num = 0
    for ch in text[zeros:]:
        digit = _INDEX.get(ch)
        if digit is None:
            raise Base58Error(f"invalid base58 character {ch!r}")
        num = num * 58 + digit
    body = num.to_bytes((num.bit_length() + 7) // 8, "big") if num else b""
    return b"\x00" * zeros + body


def multihash_sha256(data: bytes) -> bytes:
    """multihash(sha2-256, data) = 0x12 0x20 || digest (34 bytes)."""
    return b"\x12\x20" + hashlib.sha256(data).digest()


def to_multibase(data: bytes) -> str:
    """Encode bytes as multibase base58btc (the 'z' prefix)."""
    return "z" + base58btc_encode(data)


def from_multibase(text: str) -> bytes:
    """Decode a multibase 'z...' (base58btc) string."""
    if not text.startswith("z"):
        raise Base58Error("multibase: expected 'z' prefix")
    return base58btc_decode(text[1:])


def multihash_mb(data: bytes) -> str:
    """multibase(multihash(sha2-256, data))."""
    return to_multibase(multihash_sha256(data))


def cid_from_canonical(canonical: str) -> str:
    """Compute the cid over canonical bytes."""
    return multihash_mb(canonical.encode("utf-8"))


def compute_cid(value: object) -> str:
    """Canonicalize a value and return its cid."""
    return cid_from_canonical(canonicalize(value))
